### ISSUE TITLE
An error following successful statements is not reported properly

### ISSUE THREAD

**Issue by knz** [lib/pq#633]
For example, sending the string `SELECT 1; SELECT nonexistent;` in a single call to `Query()`.

The first result set reports correctly 1 row of data;

but then `HasNextResultSet` returns false, with no error.

This is incorrect; either:
- HasNextResultSet should return `true` and subsequently a method called `Err()` would report that set's error (`Err()` is a new method of the `Rows` interface in Go's `sql` package) 
- or potentially, the Next method should recognize an error following the "end of current result set" message and ensure that error is stored for reporting by the next HasNextResultSet. (When using this approach, note that Next cannot return that error itself. The error belongs to the next result set, the current one is complete without error)

Found in https://github.com/cockroachdb/cockroach/issues/16519

**Comment by mjibson** [comment]
Ugh. I found another bug when looking at this. If you close a rows before starting reading the last result set, the following query will be in a bad part of the state machine and die.

**Comment by mjibson** [comment]
Oh yes. I remember why this is hard to fix: it is a breaking change for the lib/pq API. Consider the query when a user isn't using NextResultSet (either because they are on Go < 1.8 or they just aren't using it). In that case the query would return an error. Since we don't know if a user will call NextResultSet or not, it's not obvious what we should do, but we probably shouldn't change the lib/pq API. See https://github.com/lib/pq/issues/601#issuecomment-296491023 for some more discussion on this. The solution discussed there was to add a new API that signaled to the driver to send each message type back: https://github.com/golang-sql/sqlexp/commit/c6fb99775facdd33d510f67669b3d8038014a8bc

I'm not sure how hard it would be to implement that. Certainly a few days work.

**Comment by knz** [comment]
Ack, thanks for the analysis.

What do you suggest we do for the time being?

**Comment by mjibson** [comment]
The best idea I have right now is that if this is a thing you care about, split up SQL statements instead of sending them as a single query. This is probably impossible for most lib/pq users that accept queries as user input. However in cockroach's case, since it has a built in SQL parser, this could be doable on the client side by splitting up statements at semicolons and sending them one by one.

**Comment by knz** [comment]
I'm sorry I didn't answer sooner on this. The reason why I'm not excited about your proposed workaround is that sending the statements separately prevents automatic server-side retries.

I think either extending the `lib/pq` API (and use that directly in the crdb client, instead of gosql) or adopting some other API like `sqlexp` is the way to go.

**Comment by andreimatei** [comment]
Why do you say that returning a result set if the error is part of a future result set would be a breaking change to the lib/pq API? Or, which API exactly? The golang sql package API seems to be silent on the topic.

**Comment by mjibson** [comment]
@knz I agree about the proposed workaround being bad, it's sadly the best we can do at the moment. I'll look deeper into the sqlexp stuff.

@andreimatei It's a breaking API change to lib/pq because if we make that change and users update their lib/pq version, then they would stop getting errors for things that they were previously getting errors for. lib/pq can't know if they are going to call NextResultSets or not, and so doesn't know when it should return an error if it encounters one.

**Comment by andreimatei** [comment]
I understand it'd be a behavior change, but how is it an API change? If the API we're talking about is the golang one, the topic seems underspecified. Like, perhaps more clients will be happy with the new behaviour.
One option could be exposing a per-process lib/pq settings knob which affects the behavior of the golang API.

**Comment by mjibson** [comment]
I guess I'm just using the term API change differently than you, but yes.

I like the idea of an option in the connection string that configures the behavior of this specific thing. It's explicit and won't require doing a huge amount of the sqlexp work.

**Comment by andreimatei** [comment]
It seems to me that the golang `DB.Query()` method was useless for multi-statement "queries" before Go 1.8. If it were up to me, I wouldn't try to maintain any compatibility (or at least not make that compatibility the default).
The docs for `Rows.Err()` say:
> Err returns the error, if any, that was encountered during iteration. Err may be called after an explicit or implicit Close.

To me, the "during iteration" part means that a `Rows.Next()` call returned `false` because of the error. This is not true if the error is part of a different result set.

**Comment by arp242** [comment]
It looks like this has been fixed at some point: rows.Err() returns the error. You do need to actually read the result, which seems reasonable – PostgreSQL won't send the error until we do.

Added a test for this in #1308.
